src/registry.ts

```diff
@@ -0,0 +1,139 @@
+
+  static getDefaultRegistry(): Registry | undefined {
+    return Registry.defaultRegistry;
+  }
+
+  // If you call `enableDefaultRegistry`, any new elements will auto-register with
+  // the provided registry as soon as they're constructed.
+  static enableDefaultRegistry(registry: Registry): void {
+    Registry.defaultRegistry = registry;
+  }
+
+  static disableDefaultRegistry(): void {
+    Registry.defaultRegistry = undefined;
+  }
+
+  protected index: Index;
+
+  constructor() {
+    this.index = new Index();
+  }
+
+  clear(): Registry {
+    this.index = new Index();
+    return this;
+  }
+
+  setIndexValue(name: string, value: string, id: string, elem: Element): void {
+    const index = this.index;
+    if (!index[name][value]) {
+      index[name][value] = {};
+    }
+    index[name][value][id] = elem;
+  }
+
+  // Updates the indexes for element 'id'. If an element's attribute changes
+  // from A -> B, make sure to remove the element from A.
+  updateIndex(id: string, name: string, value: string | undefined, oldValue: string | undefined): void {
+    const elem = this.getElementById(id);
+    if (oldValue !== undefined && this.index[name][oldValue]) {
+      delete this.index[name][oldValue][id];
+    }
+    if (value && elem) {
+      this.setIndexValue(name, value, elem.getAttribute('id'), elem);
+    }
+  }
+
+  // Register element `elem` with this registry. This adds the element to its index and watches
+  // it for attribute changes.
+  register(elem: Element, id?: string): Registry {
+    id = id || elem.getAttribute('id');
+    if (!id) {
+      throw new X("Can't add element without `id` attribute to registry", elem);
+    }
+
+    // Manually add id to index, then update other indexes.
+    elem.setAttribute('id', id);
+    this.setIndexValue('id', id, id, elem);
+    this.updateIndex(id, 'type', elem.getAttribute('type'), undefined);
+    elem.onRegister(this);
+    return this;
+  }
+
+  getElementById(id: string): Element | undefined {
+    if (this.index.id && this.index.id[id]) {
+      return this.index.id[id][id];
+    } else {
+      return undefined;
+    }
+  }
+
+  getElementsByAttribute(attribute: string, value: string): Element[] {
+    const index_attr = this.index[attribute];
+    if (index_attr) {
+      const index_attr_val = index_attr[value];
+      if (index_attr_val) {
+        const keys = Object.keys(index_attr_val);
+        return keys.map((k) => index_attr_val[k]);
+      }
+    }
+    return [];
+  }
+
+  getElementsByType(type: string): Element[] {
+    return this.getElementsByAttribute('type', type);
+  }
+
+  getElementsByClass(className: string): Element[] {
+    return this.getElementsByAttribute('class', className);
+  }
+
+  // This is called by the element when an attribute value changes. If an indexed
+  // attribute changes, then update the local index.
+  onUpdate(id: string, name: string, value: string | undefined, oldValue: string | undefined): Registry {
```

Comment:
This changes the interface, and can break existing uses. We should take an object parameter (like the prior implementation) for backward compatibility.

Reply:
@0xfe do you refer to `id` ?

Reply:
yes, correct -- I mean the original interface took a single object parameter.

Reply:
Will do!

I originally assumed onUpdate() was a private API, since it is only used by element.ts, and does not show up in the tests. Changing it to separate parameters with explicit types reduces the number of TypeScript warnings and required conditional checks.... **but** if it has been used by external projects and consumed as a public API, then we definitely need to not break the API.

I'll make the changes and update this PR.

Reply:
Done!